### PR TITLE
Add AWS presigned URL support

### LINKED ISSUE #4763: [object_store] Support generating and using signed upload URLs
IOx needs the ability to ask object storage for a signed URL that can be then used for a PUT request to upload a file to a particular path.

This is related to, but slightly different from https://github.com/apache/arrow-rs/issues/3027, which is for generating signed URLs for GET requests to read particular files. (it might be easy enough to add that functionality when this ticket is done.

### PR DESCRIPTION
**Pull request by carols10cents**
Ok, I'm feeling pretty good about this-- I did test it against a real live S3 bucket and I was able to upload a file with cURL using signed URLs generated by these new functions!

There's always more refactoring that could be done... the private method where I told clippy to let me have many arguments points to the worst of it.

If you take a look commit-by-commit, the last commit adds a public `signed_url` method directly to `AmazonS3`, which I think makes it super convenient, but might be more than you'd like to expose. I'm happy to back that commit out!

# Which issue does this PR close?

Closes #4763.
Connects to #3027, which is requesting signed GET URLs for all object store implementations that support it-- this PR implements support for both GET and PUT signed URLs, but only for AWS.

# Rationale for this change

Enable reusing the authorization and signing code to generate signed URLs that can be used to give permission to upload or view particular objects. 

# What changes are included in this PR?

A new public method on `AwsAuthorizer`, currently called `sign` (as opposed to the existing `authorize` method) that adds AWS signature query parameters to a specified `Url` so that it can be given to another user to authorize uploading or viewing a particular resource.

This shares a lot of the signing code with the existing `authorize` method, so I've pulled out a few private methods, but I don't think the current structure is ideal yet. Thoughts on how far to take this are very welcome, such as if there's another struct or three starting to cry to be let out of `AwsAuthorizer`, if these should be free functions, etc.

# Are there any user-facing changes?

Yes, one (or possibly two) new method(s) in the public API. See the included examples!

### CODE REVIEW

File: object_store/src/aws/credential.rs [hunk 1]
```diff
@@ -154,21 +155,140 @@ impl<'a> AwsAuthorizer<'a> {
...
+            .append_pair("X-Amz-Algorithm", ALGORITHM)
+            .append_pair(
+                "X-Amz-Credential",
+                &format!("{}/{}", self.credential.key_id, scope),
+            )
+            .append_pair("X-Amz-Date", &date.format("%Y%m%dT%H%M%SZ").to_string())
+            .append_pair("X-Amz-Expires", &expires_in.as_secs().to_string())
+            .append_pair("X-Amz-SignedHeaders", "host");
+
+        // TODO: For S3, you must include the X-Amz-Security-Token query parameter in the URL if
```
Comment (carols10cents) [inline object_store/src/aws/credential.rs:236]: WHOOPS at some point this weekend I remembered this TODO was in here... taking care of it now! 😅

Comment (carols10cents) [reply object_store/src/aws/credential.rs]: and done ✅

**Comment by carols10cents** [comment]
Hm. I'm trying to put this into use in IOx and running into a bit of a problem... ideally, we'd like to not have to change much about the rest of the codebase that uses `Arc<object_store::DynObjectStore>` pretty much everywhere, so Dom and I were thinking of defining a new trait for presigned URLs (that would be a supertrait of `object_store::ObjectStore` and the bulk import could say it needed that, `AmazonS3` could implement it, and the other object store types could implement the new trait but return an error (at some level TBD).

The problem with that I'm running into is that we're not actually using `AmazonS3`, we're using a `LimitStore` adapting an `AmazonS3`, and I can't `impl<T: TheNewTrait> TheNewTrait for `LimitStore<T>`  because `LimitStore` (or for that matter `ThrottledStore`) doesn't expose access to the inner `T`.

The solutions I'm seeing to this need to be done in `object_store`, I think:

- Add `pub fn inner(&self) -> &T` to `LimitStore` and `ThrottledStore` to enable implementing traits that delegate to these adapters' inner object stores, which could enable other object store extensions in the future
- Define this new trait I tried to write within `object_store` itself, and have all the implementers other than `AmazonS3` return an error saying it's not supported
- ??? Something else I'm not seeing?

I'm going to try going down these paths to see if I run into different issues, but I'd love thoughts on the best way to solve this from the `object_store` perspective!

**Comment by carols10cents** [comment]
> I think I might be missing some context on why the pre-signed URL functionality needs to be conflated with the ObjectStore trait.

TL;DR IOx; I'm happy to make changes to IOx to do whatever.

That's purely due to the current IOx implementation that has a `clap_blocks` function for configuring any object store and currently returns an `Arc<DynObjectStore>`, thus erasing any type information after that point.

For the bulk import functionality, what we've planned for a piece of it is a gRPC service implemented in the router that has an action that returns a pre-signed URL. Currently, the router uses the aforementioned `clap_blocks` to configure the object store and has the `Arc<DynObjectStore>` plumbed through it everywhere-- I'd like to avoid changing all that if possible, and supporting presigned URLs through other object stores if those implementations gain the signing capability in the future... does that make more sense? I am relying on the IOx context you have, let me know if I glossed over too much :)

> I would have perhaps expected a `dyn Signer` that could be passed around to the places that need it. 

Right, but who defines `trait Signer`? I think it'd be nice if it was in `object_store` such that as more object store implementations gain that functionality, users wouldn't have to add the implementations themselves. Would that be ok or do you think it doesn't belong here?

----

Do you have any general comments about this PR as-is?

**Comment by tustvold** [comment]
I need to sleep on it, but perhaps we just add it to the ObjectStore trait with a default not implemented impl, it wouldn't be the weirdest thing there - I'm looking at you append 😅

**Comment by tustvold** [comment]
Codes looks good, will continue to think about how we expose this

**Comment by carols10cents** [comment]
Ok, so here's another wrinkle. 

The URL signing API should return a full URL so that if the user wants to, they can pass the URL to something that isn't using the `object_store` crate and use it with `cURL`, `wget`, a web browser, what have you.

However, if the user _is_ doing all upload/download through the `object_store` crate, and particularly through wrappers like `LimitStore` and such, they _will_ want to send the request using the presigned URL through their `object_store`, but currently `put_request` and friends take a `Path` that gets concatenated onto the object store implementation's full URL (which wouldn't be what we'd want in this case).

So I think the `ObjectStore` trait will need methods like `fn put_presigned(&self, location: &Url, bytes: Bytes)` and `fn get_presigned(&self, location: &Url)`, and these should be on the trait so that the adapters like `LimitStore` and `ThrottledStore` can support them as well.

In IOx terms, the router will have something that can produce signed urls (configured using the code in clap_blocks, [I think I've figured out a good way to do that](https://github.com/influxdata/influxdb_iox/pull/8927)) and will pass that full URL to a bulk import tool, that will also have an object store configured using the code in clap_blocks (that currently returns an `Arc<DynObjectStore>` which is a `LimitStore` wrapping an `AmazonS3` in this case) so that we don't have to reimplement all those settings and code handling the upload request.

Does that sound right to you?

**Comment by carols10cents** [comment]
> n IOx terms, the router will have something that can produce signed urls (configured using the code in clap_blocks, https://github.com/influxdata/influxdb_iox/pull/8927) and will pass that full URL to a bulk import tool, that will also have an object store configured using the code in clap_blocks (that currently returns an Arc<DynObjectStore> which is a LimitStore wrapping an AmazonS3 in this case) so that we don't have to reimplement all those settings and code handling the upload request.

As Raphael and I were chatting, I realized this statement of mine was exactly wrong, as Raphael also commented-- the whole point of the thing using the signed URLs is that it DOESN'T have the `object_store` credentials etc configured-- everything needed will be in the URL. So yeah, ignore this-- the actual upload shouldn't happen through the `object_store` crate.

**Comment by carols10cents** [comment]
Ok, I think I'm happy-ish with this? I have [an iox branch working](https://github.com/influxdata/influxdb_iox/pull/8927) that I think isn't terrible, maybe?

CI is failing with something about ring, I don't know why :(

**Comment by tustvold** [comment]
Ring failure was likely - https://github.com/briansmith/ring/issues/1707

File: object_store/src/aws/mod.rs [hunk 2]
```diff
@@ -209,6 +210,65 @@ impl AmazonS3 {
...
+    ///
+    /// This example returns a URL that will enable a user to upload a file to
+    /// "some-folder/some-file.txt" in the next hour.
+    ///
+    /// ```
+    /// # async fn example() -> Result<(), Box<dyn std::error::Error>> {
+    /// use object_store::{aws::AmazonS3Builder, path::Path, signer::Signer};
+    /// use http::Method;
+    /// use std::time::Duration;
+    ///
```
Comment (tustvold) [inline object_store/src/aws/mod.rs:237]:
```suggestion
    /// # use object_store::{aws::AmazonS3Builder, path::Path, signer::Signer};
    /// # use http::Method;
    /// # use std::time::Duration;
    /// #
```

File: object_store/Cargo.toml [hunk 3]
```diff
@@ -45,6 +45,7 @@ walkdir = "2"
 
 # Cloud storage support
 base64 = { version = "0.21", default-features = false, features = ["std"], optional = true }
+http = { version = "0.2", default_features = false, optional = true }
```
Comment (tustvold) [inline object_store/Cargo.toml:48]: FWIW reqwest re-exports `http::Method` which appears to be the only thing this is used for

**Review by tustvold — APPROVED** [review]
Makes sense to me

**Comment by carols10cents** [comment]
@tustvold fixed, and CI is passing now!! 🎉
